### PR TITLE
Implement Multinomial distribution

### PR DESCRIPTION
**Pull request by alicanb**
cc @apaszke. This was team effort between @fritzo, @neerajprad, and me.

### CODE REVIEW

File: torch/distributions/multinomial.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+        shifted_idx = list(range(samples.dim()))
+        shifted_idx.append(shifted_idx.pop(0))
+        samples = samples.permute(*shifted_idx)
+        counts = samples.new(self._extended_shape(sample_shape)).zero_()
+        counts.scatter_add_(-1, samples, torch.ones_like(samples))
+        return counts.type_as(self.probs)
+
+    def log_prob(self, value):
+        self._validate_log_prob_arg(value)
+        logits, value = broadcast_all(self.probs.log(), value)
```
Comment (neerajprad) [inline torch/distributions/multinomial.py]: The issue here was that we need to clone the logits because they are getting mutated in line 89 below. We can change that and resort to using `self.probs_or_logits` instead as part #4614.

File: torch/distributions/multinomial.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+              'probs': constraints.simplex, 'logits': constraints.real}
+
+    def __init__(self, total_count=1, probs=None, logits=None):
+        if isinstance(total_count, Number):
+            self.total_count = total_count
+        else:
+            if isinstance(total_count, Variable):
+                total_count = total_count.data
+            self.total_count = total_count.view(-1)[0]
+            if (total_count != self.total_count).any():
```
Comment (apaszke) [inline torch/distributions/multinomial.py]: wouldn't it be simpler to disallow passing `total_count` vectors at all (just check that `total_count.numel() == 1`)?

File: torch/distributions/multinomial.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    called (see example below)
+
+    -   :meth:`sample` requires a single shared `total_count` for all
+        parameters and samples.
+    -   :meth:`log_prob` allows different `total_count` for each parameter and
+        sample.
+
+    Example::
+
+        >>> m = Multinomial(10, torch.Tensor([ 1, 1, 1, 1]))
```
Comment (apaszke) [inline torch/distributions/multinomial.py]: nit: can we increase `total_count` to something larger (e.g. 100), so that it's clear that `1, 3, 3, 3` below aren't class indices but counts?

File: torch/distributions/multinomial.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+        >>> x= m.sample()  # equal probability of 0, 1, 2, 3
```
Comment (apaszke) [inline torch/distributions/multinomial.py]: nit: space

File: torch/distributions/multinomial.py (hunk 1 above)
Comment (apaszke) [reply torch/distributions/multinomial.py]: Can't we do `self.logits.clone()`? That's likely to be faster than `log()` + if someone gives you `logits` in the constructor we'll do `logits` -> `probs` (because `self.probs` is accessed) -> `logits`, which isn't a good idea

File: torch/distributions/multinomial.py (hunk 2 above)
Comment (fritzo) [reply torch/distributions/multinomial.py]: Done.

**Review by apaszke — APPROVED** [review]

**Comment by neerajprad** [comment]
Can we get this merged into master?
